### DISCUSSION TITLE
Navigation with HJKL

### DISCUSSION THREAD

**Discussion by bennypowers** [nvim-neo-tree/neo-tree.nvim#163] [Q&A] [answered]
Seeking to emulate atom's tree-view component, I'd like for hjkl to operate like so in neo-tree buffers:

h - out (i.e up to parent dir when on a file, close dir when on a dir)
j - down (i.e. next sibling, or focus next dir when at last sibling)
k - up (i.e. prev sibling, or focus parent dir when at first sibling)
l - in (i.e. open dir when on a dir, noop when on a file)

does neotree expose commands to setup those bindings?

**Comment by cseickel** [comment, marked as answer]
`out` and `in` don't exist, but you can create custom functions that do this, see:
`:h neo-tree-custom-commands`
https://github.com/nvim-neo-tree/neo-tree.nvim/blob/75cde5dd98879b05d01514ba588d9cdb8e3171b6/lua/neo-tree/sources/common/commands.lua#L180
https://github.com/nvim-neo-tree/neo-tree.nvim/blob/75cde5dd98879b05d01514ba588d9cdb8e3171b6/lua/neo-tree/sources/filesystem/init.lua#L326

With the `next_sibling`/`prev_sibling` commands, I know I've used them before but I can't remember, in what situation is it different than a standard j/k navigation?

**Reply by bennypowers** [reply]
I think it's basically the same, except it'd be nice for `*_sibling` to place the cursor on the icon, instead of regular block movement

**Reply by bennypowers** [reply]
This is more-or-less what I wanted:

```lua
["h"] = function(state)
  local node = state.tree:get_node()
    if node.type == 'directory' and node:is_expanded() then
      require'neo-tree.sources.filesystem'.toggle_directory(state, node)
    else
      require'neo-tree.ui.renderer'.focus_node(state, node:get_parent_id())
    end
  end,
["l"] = function(state)
  local node = state.tree:get_node()
    if node.type == 'directory' then
      if not node:is_expanded() then
        require'neo-tree.sources.filesystem'.toggle_directory(state, node)
      elseif node:has_children() then
        require'neo-tree.ui.renderer'.focus_node(state, node:get_child_ids()[1])
      end
    end
  end,
```

**Reply by mehalter** [reply]
I also was interested in this but wanted the functions to work in all of the different source selectors, so I made a similar snippet that uses the `toggle_node` command rather than the `filesystem` specific `toggle_directory`. I figured I would drop it here just in case anyone else finds it useful:

```lua
h = function(state)
  local node = state.tree:get_node()
  if (node.type == "directory" or node:has_children()) and node:is_expanded() then
    state.commands.toggle_node(state)
  else
    require("neo-tree.ui.renderer").focus_node(state, node:get_parent_id())
  end
end,
l = function(state)
  local node = state.tree:get_node()
  if node.type == "directory" or node:has_children() then
    if not node:is_expanded() then
      state.commands.toggle_node(state)
    else
      require("neo-tree.ui.renderer").focus_node(state, node:get_child_ids()[1])
    end
  end
end,
```

**Comment by kaem-e** [comment]
hey is there a way to modify this to allow pressing l on a file itll just automatically decide to open instead of expanding the directory ?

**Reply by dvargas46** [reply]
I came across this while looking for the same type of functionality. Basically on top of collapsing/expanding of folders, I would like to be able to open files with "l", when selected. I was able to achieve it with:
```lua
require("neo-tree.sources.filesystem.commands").open(state)
```

on any of the two solutions provided above, for the "l" mapping you can add this in an else condition when it's checking whether the node is a directory/has children. maybe there's another way, but this ended up working for me.

**Comment by farawayisland** [comment]
For future readers who use a `lazy.nvim` setup, here's a summary and combination of both [mehalter's](https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/163#discussioncomment-4747082) and [dvargas46's](https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/163#discussioncomment-7663286) solutions which I hope might be useful:

```
return {
  'nvim-neo-tree/neo-tree.nvim',
  version = '*',
  dependencies = {
    'nvim-lua/plenary.nvim',
    'nvim-tree/nvim-web-devicons', -- optional, but recommended
    'MunifTanjim/nui.nvim',
  },
  lazy = false,
  opts = {
    sources = {
      'filesystem',
      'buffers',
      'git_status',
      'document_symbols',
    },
    source_selector = {
      winbar = true,
      sources = {
        { source = 'filesystem', display_name = ' 󰉓 File ' },
        { source = 'git_status', display_name = ' 󰊢 Git ' },
        { source = 'buffers', display_name = ' 󰓩 Buf ' },
        { source = 'document_symbols', display_name = '  Sym ' },
      },
      content_layout = 'center',
    },
    window = {
      mappings = {
        -- Sources:
        -- https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/163#discussioncomment-4747082
        -- https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/163#discussioncomment-7663286
        -- Jump up to parent directory on file or closed directory, or close on open directory
        ['h'] = function(state)
          local node = state.tree:get_node()
          if (node.type == 'directory' or node:has_children()) and node:is_expanded() then
            state.commands.toggle_node(state)
          else
            require('neo-tree.ui.renderer').focus_node(state, node:get_parent_id())
          end
        end,
        -- Open on file or closed directory, or jump down to top subdirectory on open directory
        ['l'] = function(state)
          local node = state.tree:get_node()
          if node.type == 'directory' or node:has_children() then
            if not node:is_expanded() then
              state.commands.toggle_node(state)
            else
              require('neo-tree.ui.renderer').focus_node(state, node:get_child_ids()[1])
            end
          else
            require('neo-tree.sources.filesystem.commands').open(state)
          end
        end,
        ['<S-Tab>'] = 'prev_source',
        ['<Tab>'] = 'next_source',
      },
    },
  },
  config = function(_, opts)
    require('neo-tree').setup(opts)
    -- Additional configs here
  end,
}
```
